sdk/storage/src/common/retriable_stream.cpp

```diff
@@ -0,0 +1,46 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include "common/retriable_stream.hpp"
+
+using Azure::Core::Context;
+using Azure::Core::Http::BodyStream;
+
+namespace Azure { namespace Storage {
+
+  int64_t RetriableStream::Read(Context& context, uint8_t* buffer, int64_t count)
+  {
+    if (this->m_options.DoInjectError)
+    {
+      throw std::runtime_error("do we say something here for testing?");
+    }
+
+    char const* error;
+    for (int64_t intent = 0; intent < this->m_options.MaxRetryRequests; intent++)
+    {
+      // check if we need to get inner stream
+      if (this->m_inner == nullptr)
+      {
+        // Get a new inner bodyStream that start from last offset and size m_count (remaining)
+        // if this fails, throw should bubble up
+        this->m_inner = this->m_retryFn(context, this->m_retryInfo);
```

Comment:
This will destruct the previous inner stream, right? We need to ensure that the previous stream (and its socket) gets closed. If so, you should also add comments here about this - it is very important.

Reply:
Yes, since we have the BodyStream as a unique_ptr, as soon as we reassing it, the previous one is gone (bodyStream destructor is called to clean-up libcurl staff, or any other http stack, if not libcurl)